gammapy/maps/sparse.py

```diff
@@ -0,0 +1,169 @@
+
+    @property
+    def idx(self):
+        """Return flattened index vector."""
+        return self._idx
+
+    @property
+    def shape(self):
+        """Return the array shape."""
+        return self._shape
+
+    @property
+    def ndim(self):
+        """Return the array dimension."""
+        return len(self._shape)
+
+    @classmethod
+    def from_array(cls, data, min_value=0):
+        """Create from a numpy array.
+
+        Parameters
+        ----------
+        data : `numpy.ndarray`
+            Input data array.
+
+        min_value : float
+            Threshold for sparsifying the data vector.
+        """
+        shape = data.shape
+        idx = np.where(data > min_value)
+        idx = np.ravel_multi_index(idx, shape)
+        data = data[data > min_value]
+        return cls(shape, idx, data)
+
+    def _to_flat_index(self, idx_in):
+        """Convert index tuple to a flattened index."""
+        idx_in = tuple([np.array(z, ndmin=1, copy=False) for z in idx_in])
+        msk = np.all(
+            np.stack([t < n for t, n in zip(idx_in, self.shape)]), axis=0)
+        idx = np.ravel_multi_index(tuple([t[msk] for t in idx_in]),
+                                   self.shape, mode='wrap')
+
+        return idx, msk
+
+    def set(self, idx_in, vals, fill=False):
+        """Set array values at indices ``idx_in``."""
+
+        o = np.broadcast_arrays(vals, *idx_in)
+        vals = np.ravel(o[0])
+
+        # TODO: Determine whether new vs. existing indices are being
+        # addressed, in the latter case we only need to update data
+        # array
+
+        vals = np.array(vals, ndmin=1)
+        idx_flat_in, msk_in = self._to_flat_index(idx_in)
+        idx, data = merge_sparse_arrays(idx_flat_in, vals,
+                                        self.idx, self.data,
+                                        fill=fill)
+
+        self._idx = idx
+        self._data = data
+        #idx, msk = find_in_array(idx_flat_in, self.idx)
+        #self._data[idx[msk]] = vals[msk]
+
+    def get(self, idx_in):
+        """Get array values at indices ``idx_in``."""
+
+        shape_out = idx_in[0].shape
+        size_out = idx_in[0].size
+        idx_flat_in, msk_in = self._to_flat_index(idx_in)
+
+        # Get broadcasted shape?
+
+        idx, msk = find_in_array(idx_flat_in, self.idx)
+        val_out = np.zeros(shape_out)
+        val_out.flat[np.flatnonzero(msk_in)[msk]] = self._data[idx[msk]]
+        return np.squeeze(val_out)
+
+    def sum(self, axis=None, dtype=None, out=None, keepdims=False, **unused_kwargs):
+
+        # FIXME: Figure out how to correctly support np.apply_over_axes
+        
+        if axis is None:
+            return np.sum(self._data)
+        else:
+            shape = list(self.shape)
+            if keepdims:
+                shape[axis] = 1
+            else:
+                del shape[axis]
+            out = SparseArray(shape)
+            return out
+
+    def __getitem__(self, slices):
```

Comment:
I've started to clean up Gammapy towards always putting the special dunder methods like `__getitem__` at the top, after `__init__`. I think its a pretty common Python coding pattern. Suggest to do that everywhere and move them up for this class as well.

Reply:
Sounds like a good convention.  I'll plan to implement with the other changes.